File: backend/app/schemas/survey.py
"""Survey Schemas"""
from typing import Optional, List, Dict
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum


class FilledByType(str, Enum):
    sendiri = "sendiri"
    diwakilkan = "diwakilkan"


class QuestionType(str, Enum):
    rating = "rating"
    text = "text"
    multiple_choice = "multiple_choice"


# ============ Service Type Schemas ============

class ServiceTypeCreate(BaseModel):
    """Schema for creating a service type"""
    name: str = Field(..., min_length=1, max_length=200)


class ServiceTypeUpdate(BaseModel):
    """Schema for updating a service type"""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    is_active: Optional[bool] = None


class ServiceTypeResponse(BaseModel):
    """Schema for service type response"""
    id: int
    name: str
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ServiceTypeListResponse(BaseModel):
    """Schema for service type list"""
    items: List[ServiceTypeResponse]
    total: int


# ============ Survey Question Schemas ============

class SurveyQuestionCreate(BaseModel):
    """Schema for creating a survey question"""
    question_text: str = Field(..., min_length=1)
    question_type: QuestionType = QuestionType.rating
    options: Optional[List[str]] = None
    is_required: bool = True
    order: int = 0


class SurveyQuestionUpdate(BaseModel):
    """Schema for updating a survey question"""
    question_text: Optional[str] = Field(None, min_length=1)
    question_type: Optional[QuestionType] = None
    options: Optional[List[str]] = None
    is_required: Optional[bool] = None
    is_active: Optional[bool] = None
    order: Optional[int] = None


class SurveyQuestionResponse(BaseModel):
    """Schema for survey question response"""
    id: int
    question_text: str
    question_type: QuestionType
    options: Optional[List[str]] = None
    is_required: bool
    is_active: bool
    order: int
    created_at: datetime

    class Config:
        from_attributes = True


class SurveyQuestionListResponse(BaseModel):
    """Schema for survey question list"""
    items: List[SurveyQuestionResponse]
    total: int


class ReorderQuestionsRequest(BaseModel):
    """Schema for reordering questions"""
    question_ids: List[int]


# ============ Survey Response Schemas ============

class SurveySubmit(BaseModel):
    """Schema for submitting a survey response"""
    service_type_id: int
    filled_by: FilledByType
    responses: Dict[str, str]  # {question_id: answer}
    feedback: Optional[str] = None


class SurveyResponseDetail(BaseModel):
    """Schema for survey response detail"""
    id: int
    service_type_id: int
    service_type_name: str
    filled_by: FilledByType
    responses: Dict[str, str]
    feedback: Optional[str] = None
    submitted_at: datetime

    class Config:
        from_attributes = True


class SurveyResponseListResponse(BaseModel):
    """Schema for paginated survey response list"""
    items: List[SurveyResponseDetail]
    total: int
    page: int
    per_page: int


# ============ Survey Stats Schema ============

class ServiceTypeStats(BaseModel):
    """Stats for a service type"""
    service_type_id: int
    service_type_name: str
    total: int
    rating_distribution: Dict[str, int]


class SurveyStatsResponse(BaseModel):
    """Schema for survey statistics"""
    total_responses: int
    rating_distribution: Dict[str, int]
    by_service_type: List[ServiceTypeStats]
    by_filled_by: Dict[str, int]


# ============ Per-Question Stats Schema ============

class TextFeedbackItem(BaseModel):
    """Schema for individual text feedback"""
    response_id: int
    answer: str
    service_type_name: str
    submitted_at: datetime


class QuestionStatistics(BaseModel):
    """Schema for per-question statistics"""
    question_id: int
    question_text: str
    question_type: QuestionType
    response_count: int
    rating_distribution: Optional[Dict[str, int]] = None  # For rating questions
    text_responses: Optional[List[TextFeedbackItem]] = None  # For text questions


class QuestionStatsResponse(BaseModel):
    """Schema for all question statistics"""
    questions: List[QuestionStatistics]
    total_responses: int

File: frontend/src/types/survey.ts
// Survey Kepuasan Types

// Rating levels
export type SatisfactionRating = 'sangat_puas' | 'puas' | 'cukup_puas' | 'tidak_puas' | 'sangat_tidak_puas';

export const SATISFACTION_LABELS: Record<SatisfactionRating, string> = {
  sangat_puas: 'Sangat Puas',
  puas: 'Puas',
  cukup_puas: 'Cukup Puas',
  tidak_puas: 'Tidak Puas',
  sangat_tidak_puas: 'Sangat Tidak Puas',
};

export const SATISFACTION_ICONS: Record<SatisfactionRating, string> = {
  sangat_puas: '😊',
  puas: '🙂',
  cukup_puas: '😐',
  tidak_puas: '☹️',
  sangat_tidak_puas: '😢',
};

export const SATISFACTION_COLORS: Record<SatisfactionRating, string> = {
  sangat_puas: 'bg-green-500',
  puas: 'bg-blue-500',
  cukup_puas: 'bg-yellow-500',
  tidak_puas: 'bg-orange-500',
  sangat_tidak_puas: 'bg-red-500',
};

// Color hex untuk SVG (karena SVG tidak support Tailwind classes)
export const SATISFACTION_COLOR_HEX: Record<SatisfactionRating, string> = {
  sangat_puas: '#22c55e',
  puas: '#3b82f6',
  cukup_puas: '#eab308',
  tidak_puas: '#f97316',
  sangat_tidak_puas: '#ef4444',
};

// Jenis Layanan
export interface ServiceType {
  id: number;
  name: string;
  is_active: boolean;
  created_at: string;
}

// Survey Question
export type QuestionType = 'rating' | 'text' | 'multiple_choice';

export interface SurveyQuestion {
  id: number;
  question_text: string;
  question_type: QuestionType;
  options?: string[]; // For multiple choice
  is_required: boolean;
  is_active: boolean;
  order: number;
  created_at: string;
}

// Survey Response
export interface SurveyResponse {
  id: number;
  service_type_id: number;
  service_type_name?: string;
  filled_by: 'sendiri' | 'diwakilkan';
  responses: Record<number, string | SatisfactionRating>; // question_id -> answer
  feedback?: string;
  submitted_at: string;
}

// Form data for submitting survey
export interface SurveyFormData {
  service_type_id: number;
  filled_by: 'sendiri' | 'diwakilkan';
  responses: Record<number, string | SatisfactionRating>;
  feedback?: string;
}

// Survey Statistics
export interface SurveyStats {
  total_responses: number;
  rating_distribution: Record<SatisfactionRating, number>;
  by_service_type: {
    service_type_id: number;
    service_type_name: string;
    total: number;
    rating_distribution: Record<SatisfactionRating, number>;
  }[];
  by_filled_by: {
    sendiri: number;
    diwakilkan: number;
  };
}

// Backend response types
export interface BackendServiceType {
  id: number;
  name: string;
  is_active: boolean;
  created_at: string;
}

export interface BackendSurveyQuestion {
  id: number;
  question_text: string;
  question_type: QuestionType;
  options: string[] | null;
  is_required: boolean;
  is_active: boolean;
  order: number;
  created_at: string;
}

export interface BackendSurveyResponse {
  id: number;
  service_type_id: number;
  service_type_name: string;
  filled_by: 'sendiri' | 'diwakilkan';
  responses: Record<number, string>;
  feedback: string | null;
  submitted_at: string;
}

export interface BackendSurveyResponseList {
  items: BackendSurveyResponse[];
  total: number;
  page: number;
  per_page: number;
}

export interface BackendSurveyStats {
  total_responses: number;
  rating_distribution: Record<string, number>;
  by_service_type: {
    service_type_id: number;
    service_type_name: string;
    total: number;
    rating_distribution: Record<string, number>;
  }[];
  by_filled_by: {
    sendiri: number;
    diwakilkan: number;
  };
}

// Per-Question Stats Types
export interface TextFeedbackItem {
  response_id: number;
  answer: string;
  service_type_name: string;
  submitted_at: string;
}

export interface QuestionStatistics {
  question_id: number;
  question_text: string;
  question_type: QuestionType;
  response_count: number;
  rating_distribution?: Record<SatisfactionRating, number>;
  text_responses?: TextFeedbackItem[];
}

export interface BackendQuestionStatsResponse {
  questions: QuestionStatistics[];
  total_responses: number;
}
